[Feature] Data Validation in the scores table

cprbucat2 opened the pull request:
## Old behavior
Had zero data validation or key checks, anything could be inputted.

## New behavior
Added new data validation file.
- limit input to numbers, one decimal, tab, backspace, and side arrows
- cell outlines red when text inside is invalid
- added validity checks
- [x] Block submission of invalid input.

## Additional info (related issues, images, etc.)
Closes #37.

cprbucat2 committed:
Data Validation in the score table

Added new data validation file.
- limit input to numbers, one decimal, tab, backspace, and side arrows
- cell turns red when text inside is invalid
- added validity checks

cprbucat2 committed:
Merge branch 'main' into cprbucat2/issue37

Signed-off-by: Aidan Hoover <96591790+cprbucat2@users.noreply.github.com>

cprbucat2 committed:
[Fix] Changed class names from merge

Changed in data_validation_up
Changed in data_validation_down

bobpaw commented:
We should check [`KeyboardEvent.getModifierState()`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/getModifierState) because this is preventing me from pressing Ctrl-R to reload and also blocking copy-pasting. However, I can still right click to paste invalid characters.

bobpaw commented:
https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/beforeinput_event
https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/input_event
Also do the valid input check at page load because there can be saved form content (after reload for example)

cprbucat2 committed:
Adds submission validation, Switch to input event

-changed from keyup to input events
-added validation on comment boxes (need text entered to be valid)
-added invalid comment box red outline
-added validation checks to prevent submission in submit.js
-on submit attempt any invalid boxes highlight red

bobpaw committed:
Merge branch 'main' into cprbucat2/issue37

bobpaw committed:
Move datavalidation.js

bobpaw committed:
Match resource order before merge

cprbucat2 committed:
Merge branch 'main' into cprbucat2/issue37

bobpaw commented:
@cprbucat2, Nicer inset on rows:
```diff
diff --git a/app/www/public/css/styles.css b/app/www/public/css/styles.css
index ddd284a..4817448 100644
--- a/app/www/public/css/styles.css
+++ b/app/www/public/css/styles.css
@@ -147,12 +147,14 @@ table {
 	padding-bottom: 3px;
 }
 
-.feedback-comments__member-comments--invalid {
+.feedback-comments__member td {
+	padding: 5px;
+}
+
+.feedback-comments__member--invalid {
 	background-clip: content-box;
-	box-shadow: 0 0 0 1.5px #ff0000a0;
+	box-shadow: inset 0 0 0 5px #ff0000a0;
 	margin-right: 5px;
-	padding-left: 0px;
-	padding-bottom: 5px;
 }
 
 .feedback-comments__member-comments {
diff --git a/app/www/public/js/submit.js b/app/www/public/js/submit.js
index 56778e0..96a9d92 100644
--- a/app/www/public/js/submit.js
+++ b/app/www/public/js/submit.js
@@ -76,7 +76,7 @@ function validate_comments() {
 		data = cell.value;
 		if (data=="" || typeof data=="undefined") {
 			validated = false;
-			cell.parentElement.classList.add("feedback-comments__member-comments--invalid");
+			cell.parentElement.parentElement.parentElement.classList.add("feedback-comments__member--invalid");
 		}
 	}
 	if (!validated) {
```
If you change .feedback-comments to divs it may look a little different.

cprbucat2 committed:
Error and bug fixes

- Fixes failures of js lint check.
- Fixes error in cell validation where all cells show up as invalid
even while they are valid.

cprbucat2 committed:
Make inset nicer on comment rows

Changes invalid outline size to cover the full table cell

bobpaw reviewed (approved):
Some questionable design choices, viz having the same functions declared twice in datavalidation.js and submit.js. But whatever.
